File: DeskBuilder/src/main/java/com/ucg/desk/BuilderActivity.java
package com.ucg.desk;

import android.app.Activity;
import android.graphics.Point;
import android.os.Bundle;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.GridLayout;
import android.widget.GridView;
import android.widget.ImageView;

import com.squareup.picasso.Picasso;
import com.ucg.desk.view.CardAdapter;
import com.ucg.desk.view.FullCardAdapter;

public class BuilderActivity extends Activity {

    private int mColumnWidth;

    private View.OnClickListener mListener = new View.OnClickListener() {
        @Override
        public void onClick(final View v) {
            if (v.getParent() != null) {
                ((ViewGroup)v.getParent()).removeView(v);
            }
        }
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_fullscreen);

        // Upon interacting with UI controls, delay any scheduled hide()
        // operations to prevent the jarring behavior of controls going away
        // while interacting with the UI.
        initGrids();
    }

    private void initGrids() {
        Point point = new Point();
        getWindow().getWindowManager().getDefaultDisplay().getSize(point);
        mColumnWidth = point.x / 4;
        final GridLayout desk = (GridLayout) findViewById(R.id.grid_cards_desk);
        desk.setColumnCount(2);

        final GridView cards = (GridView) findViewById(R.id.grid_cards_all);
        cards.setColumnWidth(mColumnWidth);
        cards.setAdapter(new FullCardAdapter(this, getIntent().getStringExtra("CLASS")));
        cards.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(final AdapterView<?> adapterView, final View view, final int i, final long l) {
                Card s = ((CardAdapter) cards.getAdapter()).getCards().get(i);
                addCardToDesk(s, desk);
            }
        });
    }

    private void addCardToDesk(final Card s, final GridLayout desk) {
        final ImageView imageView = new ImageView(this);
        imageView.setOnClickListener(mListener);
        imageView.setMinimumWidth(mColumnWidth);
        imageView.setMaxWidth(mColumnWidth);
        Picasso.with(this).load(s.getUri()).placeholder(R.drawable.ic_launcher).into(imageView);

        desk.addView(imageView);
    }

}

File: DeskBuilder/src/main/scala/com/ucg/desk/BuilderActivity.scala
package com.ucg.desk

import android.app.{LoaderManager, Activity}
import android.os.Bundle
import android.graphics.Point
import android.widget._
import com.ucg.desk.view.{CardAdapter, FullCardAdapter}
import android.view.{MenuItem, Menu, ViewGroup, View}
import com.squareup.picasso.Picasso
import org.scaloid.common._
import scala.collection.mutable.ArrayBuffer
import android.content.{CursorLoader, Loader, DialogInterface, ContentValues}
import android.text.TextUtils
import android.database.Cursor
import com.ucg.desk.DeskSchema.Columns._

class BuilderActivity extends Activity with SActivity with Logger with LoaderManager.LoaderCallbacks[Cursor] {

  private val cards = new ArrayBuffer[Card]
  private var kind: String = _

  private var mColumnWidth: Int = 0
  private val mListener = new View.OnClickListener {
    def onClick(v: View) {
      if (v.getParent != null) {
        v.getParent.asInstanceOf[ViewGroup].removeView(v)
        val tag = v.getTag
        if (tag != null && tag.isInstanceOf[Card]) {
          cards -= tag.asInstanceOf[Card]
          toast(cards.size.toString)
        }
      }
    }
  }

  override def onCreateLoader(id: Int, args: Bundle): Loader[Cursor] = {
    new CursorLoader(this, DeskSchema.URI, null, NAME + "=?", Array(args.getString("NAME")), null)
  }

  override def onLoadFinished(loader: Loader[Cursor], data: Cursor) {
    if (data != null && data.moveToFirst()) while (!data.isAfterLast) {
      addCardToDesk(new Card(data.getString(data.getColumnIndex(CARD_PATH))), findViewById(R.id.grid_cards_desk).asInstanceOf[GridLayout])
      data.moveToNext()
    }
    data.close()
  }

  override def onLoaderReset(loader: Loader[Cursor]): Unit = {
  }

  override def onCreate(savedInstanceState: Bundle): Unit = {
    super.onCreate(savedInstanceState)
    setContentView(R.layout.activity_fullscreen)
    initGrids()

  }

  private def initGrids() {
    val point: Point = new Point
    getWindow.getWindowManager.getDefaultDisplay.getSize(point)
    mColumnWidth = point.x / 4
    kind = getIntent.getStringExtra("CLASS")
    if (getIntent.getExtras.containsKey("NAME")) {
      getLoaderManager.initLoader(R.id.desk_loader, getIntent.getExtras, this)
    }
    val desk: GridLayout = findViewById(R.id.grid_cards_desk).asInstanceOf[GridLayout]
    desk.setColumnCount(2)
    val cards: GridView = findViewById(R.id.grid_cards_all).asInstanceOf[GridView]
    cards.setColumnWidth(mColumnWidth)
    cards.setAdapter(new FullCardAdapter(this, kind))
    cards.setOnItemClickListener(new AdapterView.OnItemClickListener {
      def onItemClick(adapterView: AdapterView[_], view: View, i: Int, l: Long) {
        val s: Card = cards.getAdapter.asInstanceOf[CardAdapter].getCards.get(i)
        addCardToDesk(s, desk)
      }
    })
  }

  private def addCardToDesk(s: Card, desk: GridLayout) {
    val imageView: ImageView = new ImageView(this)
    imageView.setOnClickListener(mListener)
    imageView.setMinimumWidth(mColumnWidth)
    imageView.setMaxWidth(mColumnWidth)
    imageView.setTag(s)
    Picasso.`with`(this).load(s.getUri).placeholder(R.drawable.ic_launcher).into(imageView)
    desk.addView(imageView)
    cards += s
  }

  override def onCreateOptionsMenu(menu: Menu): Boolean = {
    getMenuInflater.inflate(R.menu.save, menu)
    super.onCreateOptionsMenu(menu)
  }

  def saveDesk(item: MenuItem) {
    toast("Save")
    new AlertDialogBuilder("Enter desk name", "") {
      val edit = new SEditText
      setView(edit)
      neutralButton(android.R.string.ok, (d: DialogInterface, i: Int) => {
        val name = edit.getText.toString
        if (!TextUtils.isEmpty(name)) {
          val arrCV = new ArrayBuffer[ContentValues](cards.size)
          for (card <- cards) {
            import DeskSchema.Columns._
            val cv = new ContentValues(3)
            cv.put(CARD_PATH, card.path)
            cv.put(NAME, name)
            cv.put(KIND, kind)
            arrCV += cv
          }
          val num = getContentResolver.bulkInsert(DeskSchema.URI, arrCV.toArray[ContentValues])
          toast(num.toString)
          d.dismiss()
        } else {
          toast("enter name")
        }
      })
    }.show()

  }

}